File: apps/dsa-web/src/api/researchRadar.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export type ResearchRadarItem = {
  symbol?: string | null;
  ticker?: string | null;
  priority?: string | null;
  researchBias?: string | null;
  driverScores?: Record<string, number>;
  whyOnRadar?: string[];
  whatToVerify?: string[];
  invalidationObservations?: string[];
  riskFlags?: string[];
  evidenceQuality?: {
    status?: string | null;
    score?: number | null;
  } | null;
};

export type ResearchRadarOnboardingGuidance = {
  title?: string | null;
  summary?: string | null;
  conditionsDetected?: string[];
};

export type ResearchRadarEmptyStateAction = {
  label?: string | null;
  route?: string | null;
  description?: string | null;
};

export type ResearchRadarSuggestedResearchEntrypoint = {
  surface?: string | null;
  route?: string | null;
  description?: string | null;
};

export type ResearchRadarResponse = {
  schemaVersion: string;
  generatedAt?: string | null;
  researchQueue: ResearchRadarItem[];
  aggregateSummary: {
    queueQuality?: string | null;
    priorityCounts?: Record<string, number>;
    source?: {
      scannerRunId?: number | null;
      market?: string | null;
      profile?: string | null;
    } | null;
  };
  evidenceGaps: string[];
  marketContextFit?: string | null;
  noAdviceDisclosure?: string | null;
  dataQuality?: {
    status?: string | null;
    missingEvidence?: string[];
  } | null;
  onboardingGuidance?: ResearchRadarOnboardingGuidance | null;
  emptyStateActions: ResearchRadarEmptyStateAction[];
  starterResearchWorkflow: string[];
  firstRunChecklist: string[];
  suggestedResearchEntrypoints: ResearchRadarSuggestedResearchEntrypoint[];
};

function normalizeStringList(payload: unknown): string[] {
  return Array.isArray(payload)
    ? payload.map((item) => String(item ?? '').trim()).filter(Boolean)
    : [];
}

function normalizeResearchRadarResponse(payload: unknown): ResearchRadarResponse {
  const normalized = toCamelCase<ResearchRadarResponse>(payload);
  return {
    schemaVersion: normalized.schemaVersion,
    generatedAt: normalized.generatedAt ?? null,
    researchQueue: Array.isArray(normalized.researchQueue) ? normalized.researchQueue : [],
    aggregateSummary: {
      queueQuality: normalized.aggregateSummary?.queueQuality ?? null,
      priorityCounts: normalized.aggregateSummary?.priorityCounts ?? {},
      source: normalized.aggregateSummary?.source ?? null,
    },
    evidenceGaps: normalized.evidenceGaps ?? [],
    marketContextFit: normalized.marketContextFit ?? null,
    noAdviceDisclosure: normalized.noAdviceDisclosure ?? null,
    dataQuality: normalized.dataQuality ?? null,
    onboardingGuidance: normalized.onboardingGuidance ? {
      title: normalized.onboardingGuidance.title ?? null,
      summary: normalized.onboardingGuidance.summary ?? null,
      conditionsDetected: normalizeStringList(normalized.onboardingGuidance.conditionsDetected),
    } : null,
    emptyStateActions: Array.isArray(normalized.emptyStateActions)
      ? normalized.emptyStateActions.map((item) => ({
        label: item?.label ?? null,
        route: item?.route ?? null,
        description: item?.description ?? null,
      }))
      : [],
    starterResearchWorkflow: normalizeStringList(normalized.starterResearchWorkflow),
    firstRunChecklist: normalizeStringList(normalized.firstRunChecklist),
    suggestedResearchEntrypoints: Array.isArray(normalized.suggestedResearchEntrypoints)
      ? normalized.suggestedResearchEntrypoints.map((item) => ({
        surface: item?.surface ?? null,
        route: item?.route ?? null,
        description: item?.description ?? null,
      }))
      : [],
  };
}

export const researchRadarApi = {
  async getResearchRadar(params: {
    market?: string;
    profile?: string;
    limit?: number;
  } = {}): Promise<ResearchRadarResponse> {
    const response = await apiClient.get('/api/v1/research/radar', { params });
    return normalizeResearchRadarResponse(response.data);
  },
};

File: api/v1/schemas/research_radar.py
# -*- coding: utf-8 -*-
"""Typed response models for the Research Radar API."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


RESEARCH_RADAR_RESPONSE_SCHEMA_VERSION = "research_radar_api_v1"


class _ResearchRadarModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ResearchRadarConsumerIssueResponse(_ResearchRadarModel):
    label: str
    message: str
    severity: str
    category: str


class ResearchRadarDrilldownTargetResponse(_ResearchRadarModel):
    label: str
    route: str
    reason: str | None = None


class ResearchRadarOnboardingGuidanceResponse(_ResearchRadarModel):
    title: str
    summary: str
    conditionsDetected: list[str] = Field(default_factory=list)


class ResearchRadarEmptyStateActionResponse(_ResearchRadarModel):
    label: str
    route: str
    description: str


class ResearchRadarSuggestedResearchEntrypointResponse(_ResearchRadarModel):
    surface: str
    route: str
    description: str


class ResearchRadarEvidenceQualityResponse(_ResearchRadarModel):
    status: str
    score: int | float | None = None
    missingEvidence: list[str] = Field(default_factory=list)
    missingEvidenceRaw: list[str] = Field(default_factory=list)


class ResearchRadarEvidenceHubItemResponse(_ResearchRadarModel):
    key: str
    label: str
    status: str
    summary: str
    blocker: str | None = None
    nextDataAction: str
    evidenceCount: int = 0
    totalCount: int = 0
    symbols: list[str] = Field(default_factory=list)
    details: list[str] = Field(default_factory=list)
    observationOnly: Literal[True] = True
    decisionGrade: Literal[False] = False


class ResearchRadarEvidenceHubResponse(_ResearchRadarModel):
    scannerCandidates: ResearchRadarEvidenceHubItemResponse
    backtestSamples: ResearchRadarEvidenceHubItemResponse
    stockReadiness: ResearchRadarEvidenceHubItemResponse
    dataActivation: ResearchRadarEvidenceHubItemResponse
    missingEvidenceStates: list[ResearchRadarEvidenceHubItemResponse] = Field(default_factory=list)


class ResearchRadarMarketLevelFallbackCardResponse(_ResearchRadarModel):
    cardId: str
    title: str
    status: str
    severity: str
    headline: str
    reasons: list[str] = Field(default_factory=list)
    observationOnly: Literal[True] = True
    decisionGrade: Literal[False] = False


class ResearchRadarMarketLevelFallbackResponse(_ResearchRadarModel):
    available: Literal[True] = True
    label: str
    summary: str
    candidateGenerationExecuted: bool = False
    candidateUnavailableReason: str
    regime: dict[str, Any] = Field(default_factory=dict)
    productSummary: str
    evidenceCards: list[ResearchRadarMarketLevelFallbackCardResponse] = Field(default_factory=list)
    dataQuality: dict[str, Any] = Field(default_factory=dict)
    readiness: dict[str, Any] = Field(default_factory=dict)
    missingDataFamilies: list[str] = Field(default_factory=list)
    blockedProductSurfaces: list[str] = Field(default_factory=list)
    nextOperatorAction: str
    observationOnly: Literal[True] = True
    decisionGrade: Literal[False] = False


class ResearchRadarQueueItemResponse(_ResearchRadarModel):
    symbol: str
    ticker: str
    priority: Literal["high", "medium", "low"]
    researchBias: str
    researchBiasRaw: str
    researchBiasLabel: str
    researchBiasMessage: str
    driverScores: dict[str, int | float] = Field(default_factory=dict)
    whyOnRadar: list[str] = Field(default_factory=list)
    whatToVerify: list[str] = Field(default_factory=list)
    whyNotHigherPriority: list[str] = Field(default_factory=list)
    evidenceGaps: list[str] = Field(default_factory=list)
    evidenceGapsRaw: list[str] = Field(default_factory=list)
    consumerEvidenceGaps: list[ResearchRadarConsumerIssueResponse] = Field(default_factory=list)
    invalidationObservations: list[str] = Field(default_factory=list)
    duplicateEvidenceMerged: int = 0
    riskFlags: list[str] = Field(default_factory=list)
    riskFlagsRaw: list[str] = Field(default_factory=list)
    riskFlagLabels: list[str] = Field(default_factory=list)
    evidenceQuality: ResearchRadarEvidenceQualityResponse
    consumerIssues: list[ResearchRadarConsumerIssueResponse] = Field(default_factory=list)
    drilldownTargets: list[ResearchRadarDrilldownTargetResponse] = Field(default_factory=list)
    noAdviceDisclosure: str
    observationOnly: Literal[True] = True
    decisionGrade: Literal[False] = False


class ResearchRadarAggregateSummaryResponse(_ResearchRadarModel):
    candidateCount: int = 0
    queueCount: int = 0
    priorityCounts: dict[str, int] = Field(default_factory=dict)
    dominantThemes: list[str] = Field(default_factory=list)
    queueQuality: str
    duplicateEvidenceMerged: int = 0
    queueDiversity: dict[str, Any] = Field(default_factory=dict)
    source: dict[str, Any] = Field(default_factory=dict)


class ResearchRadarDataQualityResponse(_ResearchRadarModel):
    status: str
    availableCandidateCount: int = 0
    reliableCandidateCount: int = 0
    missingEvidence: list[str] = Field(default_factory=list)
    missingEvidenceRaw: list[str] = Field(default_factory=list)
    consumerIssues: list[ResearchRadarConsumerIssueResponse] = Field(default_factory=list)


class ResearchRadarResponse(_ResearchRadarModel):
    schemaVersion: Literal["research_radar_api_v1"] = RESEARCH_RADAR_RESPONSE_SCHEMA_VERSION
    generatedAt: str
    researchQueue: list[ResearchRadarQueueItemResponse] = Field(default_factory=list)
    aggregateSummary: ResearchRadarAggregateSummaryResponse
    evidenceGaps: list[str] = Field(default_factory=list)
    evidenceGapsRaw: list[str] = Field(default_factory=list)
    marketContextFit: str
    drilldownTargets: list[ResearchRadarDrilldownTargetResponse] = Field(default_factory=list)
    consumerIssues: list[ResearchRadarConsumerIssueResponse] = Field(default_factory=list)
    onboardingGuidance: ResearchRadarOnboardingGuidanceResponse | None = None
    emptyStateActions: list[ResearchRadarEmptyStateActionResponse] = Field(default_factory=list)
    starterResearchWorkflow: list[str] = Field(default_factory=list)
    firstRunChecklist: list[str] = Field(default_factory=list)
    suggestedResearchEntrypoints: list[ResearchRadarSuggestedResearchEntrypointResponse] = Field(
        default_factory=list
    )
    noAdviceDisclosure: str
    dataQuality: ResearchRadarDataQualityResponse
    evidenceHub: ResearchRadarEvidenceHubResponse
    marketLevelFallback: ResearchRadarMarketLevelFallbackResponse | None = None
    observationOnly: Literal[True] = True
    decisionGrade: Literal[False] = False


__all__ = [
    "RESEARCH_RADAR_RESPONSE_SCHEMA_VERSION",
    "ResearchRadarAggregateSummaryResponse",
    "ResearchRadarConsumerIssueResponse",
    "ResearchRadarDataQualityResponse",
    "ResearchRadarDrilldownTargetResponse",
    "ResearchRadarEvidenceHubItemResponse",
    "ResearchRadarEvidenceHubResponse",
    "ResearchRadarEvidenceQualityResponse",
    "ResearchRadarEmptyStateActionResponse",
    "ResearchRadarMarketLevelFallbackCardResponse",
    "ResearchRadarMarketLevelFallbackResponse",
    "ResearchRadarOnboardingGuidanceResponse",
    "ResearchRadarQueueItemResponse",
    "ResearchRadarResponse",
    "ResearchRadarSuggestedResearchEntrypointResponse",
]
